manta-pay/src/crypto/hash/poseidon.rs

```diff
@@ -188,8 +235,8 @@ where
     #[inline]
     fn first_round(&self, input: [&S::Field; ARITY], compiler: &mut COM) -> State<S, COM> {
```

Comment:
If `ARITY` is associated constant, we have use slice here. Also, `[&S::Field; ARITY]` does not make sense to me. It should either be `&[S::Field; ARITY]` (if we use ARITY as const generics), or `&[S::Field]` and check arity on runtime (if we use ARITY as associated type constant)

Reply:
The `[&S::Field; ARITY]` type means that we are constraining the input to be `ARITY`-many references rather than `&[S::Field; ARITY]` which is a reference to an array of values. The first one is more conservative because the values can come from anywhere, we just point to them, whereas the second one requires that the values are all packed contiguously and if they are not, it requires moving the values first before passing them into the function.

Reply:
oh that makes sense, you are right